### PR TITLE
Removed unused imports; use _ for unused vars (Pylint).

### PR DESCRIPTION
**Pull request by hjoliver**
Clean up minor issues highlight by Pylint on codacy.com.

### CODE REVIEW

**Comment by hjoliver** [comment]
(will be some minor conflicts with #2312)

File: lib/cylc/config.py [hunk 1]
```diff
@@ -632,7 +632,8 @@ def __init__(self, suite, fpath, template_vars=None,
                 self.cfg['visualization']['node attributes'].items()):
             for attr in attrs:
                 try:
-                    key, value = re.split('\s*=\s*', attr)
+                    # Check form is 'name = attr'.
+                    _, _ = re.split('\s*=\s*', attr)
```
Comment (matthewrmshin) [inline lib/cylc/config.py]: Changes in this module will likely conflict with #2312, as I have worked on them as well.

File: lib/cylc/cylc_xdot.py [hunk 2]
```diff
@@ -403,7 +400,7 @@ def get_graph(self, group_nodes=[], ungroup_nodes=[],
         graph.graph_attr['rankdir'] = self.orientation
 
         for node in graph.nodes():
-            name, point_string = TaskID.split(node.get_name())
+            name, _ = TaskID.split(node.get_name())
```
Comment (matthewrmshin) [inline lib/cylc/cylc_xdot.py]: Alternatively, we could do...

```python
name = TaskID.split(node.get_name())[0]
```

File: lib/cylc/graphing.py [hunk 3]
```diff
@@ -195,7 +194,7 @@ def cylc_remove_nodes_from(self, nodes):
         # Create a new node name for the group.
         names = set()
         index = -1
-        for group, group_nodes in sorted(groups.items()):
+        for group, _ in sorted(groups.items()):
```
Comment (matthewrmshin) [inline lib/cylc/graphing.py]: Cheaper to just iterate the dict.

```python
for group in sorted(groups):
```

File: lib/cylc/gui/app_gcylc.py [hunk 4]
```diff
@@ -91,9 +90,9 @@ def run_get_stdout(command, filter=False):
             warning_dialog(
                 "ERROR: command failed %d\n%s" % (res, err)).warn()
             return (False, [])
-    except OSError, e:
+    except OSError as exc:
         warning_dialog(
-            "ERROR: command invocation failed %s\n%s" % (str(e), err)).warn()
+            "ERROR: command invocation failed %s\n%s" % (str(exc), err)).warn()
```
Comment (matthewrmshin) [inline lib/cylc/gui/app_gcylc.py]: No need to stringify `exc`. The template will do it automatically.

File: lib/cylc/config.py (hunk 1 above)
Comment (hjoliver) [inline lib/cylc/config.py]: Yes, unfortuantely I noticed that (see my comment above) after doing this.

File: lib/cylc/cylc_xdot.py (hunk 2 above)
Comment (hjoliver) [inline lib/cylc/cylc_xdot.py]: True, but does it matter one way or the other?

File: lib/cylc/graphing.py (hunk 3 above)
Comment (hjoliver) [inline lib/cylc/graphing.py]: That's a good point.

File: lib/cylc/gui/app_gcylc.py (hunk 4 above)
Comment (hjoliver) [inline lib/cylc/gui/app_gcylc.py]: Yep, I'll fix that too.

File: lib/cylc/cylc_xdot.py (hunk 2 above)
Comment (hjoliver) [inline lib/cylc/cylc_xdot.py]: (I'll do this on the basis that it just looks nicer to not have ignored/dummy variables).

File: lib/cylc/config.py [hunk 5]
```diff
@@ -651,7 +652,7 @@ def __init__(self, suite, fpath, template_vars=None,
         # require reordering task_attr in lib/cylc/graphing.py).
 
         self.leaves = self.get_task_name_list()
-        for ns, ancestors in self.runtime['first-parent ancestors'].items():
+        for _, ancestors in self.runtime['first-parent ancestors'].items():
```
Comment (hjoliver) [inline lib/cylc/config.py]: yes, I should have looked more carefully (I was just blindly changing unused variables to `_`!)

File: lib/cylc/config.py (hunk 1 above)
Comment (hjoliver) [inline lib/cylc/config.py]: > re.match('\s*=\s*', attr)?

Has to be `re.search` or else include initial word chars.  I'll just count `=` chars.

File: lib/cylc/dump.py [hunk 6]
```diff
@@ -32,8 +31,6 @@ def get_stop_state_summary(lines):
     for line in list(lines):
         if line.startswith('Remote command'):
             lines.remove(line)
-    line0 = lines.pop(0)
```
Comment (hjoliver) [inline lib/cylc/dump.py]: Apparently not! (unused code).

File: lib/cylc/gui/dbchooser.py [hunk 7]
```diff
@@ -156,11 +155,10 @@ def my_get_iter(item):
         # new items at this level
         new_items = new.keys()
         old_items = []
-        prune = []
 
         while iter_:
             # iterate through old items at this level
-            item, state, descr, dir = ts.get(iter_, 0, 1, 2, 3)
+            item, state, _, dir = ts.get(iter_, 0, 1, 2, 3)
```
Comment (hjoliver) [inline lib/cylc/gui/dbchooser.py]: Sensible.

File: lib/cylc/which.py [hunk 8]
```diff
@@ -1,38 +0,0 @@
...
-#
-# This program is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU General Public License for more details.
-#
-# You should have received a copy of the GNU General Public License
-# along with this program.  If not, see <http://www.gnu.org/licenses/>.
-
-# python implementation of the unix 'which' command
```
Comment (hjoliver) [inline lib/cylc/which.py]: 😬

**Comment by hjoliver** [comment]
rebased to master ... let's see if it worked (conflicts)

File: lib/cylc/config.py [hunk 9]
```diff
@@ -1456,9 +1457,10 @@ def generate_taskdefs(self, orig_expr, left_nodes, right, seq):
             try:
                 name, offset_is_from_icp, _, offset, _ = (
                     GraphNodeParser.get_inst().parse(node))
-            except GraphNodeError, x:
+            except GraphNodeError as exc:
+>>>>>>> Removed unused imports; use _ for unused vars (Pylint).
```
Comment (oliver-sanders) [inline lib/cylc/config.py]: This would appear to be the cause of the test battery failure!

Comment (hjoliver) [inline lib/cylc/config.py]: Ah, man.  I tried to do the world's fastest conflict resolution on the way out of the office.  Spectacular fail.

**Comment by hjoliver** [comment]
Note I've dealt with most of the remaining dummy assignments.

File: lib/cylc/dump.py (hunk 6 above)
Comment (matthewrmshin) [inline lib/cylc/dump.py]: Although `line0` is not used, `lines` is still used below, so we'll still need to remove the first element from it.

File: lib/cylc/gui/app_gcylc.py [hunk 10]
```diff
@@ -656,10 +646,9 @@ def setup_views(self):
         num_views = 2
         self.view_containers = []
         self.current_view_toolitems = []
-        for i in range(num_views):
-            self.current_views.append(None)
-            self.view_containers.append(gtk.HBox())
-            self.current_view_toolitems.append([])
+        self.current_views += [None] * num_views
+        self.view_containers += [gtk.HBox()] * num_views
+        self.current_view_toolitems += [[]] * 3
```
Comment (matthewrmshin) [inline lib/cylc/gui/app_gcylc.py]: Why 3?

Also:
* Can we combine lines 647-648 with lines 650 and 651?
* If all of these just need a `* 2`, we can get rid of `num_views` as well.

Comment (hjoliver) [inline lib/cylc/gui/app_gcylc.py]: Argh - I'm on a roll this week!   No idea how I did that (I blame rushed trip prep) - currently transiting Melbourne airport, I'll fix this in the Northern Hemisphere...)

**Comment by matthewrmshin** [comment]
(Doubly approved, but unable to merge due to conflicts.)

**Comment by oliver-sanders** [comment]
Profiling results for the complex suite:

| Version                                   | Run            | Elapsed Time (s)  | CPU Time - Total (s)  | Max Memory (kb) |
| --- | --- | --- | --- | --- |
| b120ff3f232de5cad59b1ca0ec5e955d388aef54  | complex suite  | 1035.8            | 309.3                 | 75412.0 |       
| hjoliver/codacy-unused-code               | complex suite  | 1030.8           | 305.7                | 75072.0    |

File: lib/cylc/dump.py (hunk 6 above)
Comment (oliver-sanders) [inline lib/cylc/dump.py]: Taking a better look at the following lines this seems OK as the code scans through lines looking for the pattern `time :`.

```python
line1 = lines.pop(0)
while not line1.startswith("time :"):
    line1 = lines.pop(0)
```

File: lib/cylc/gui/app_gcylc.py (hunk 10 above)
Comment (hjoliver) [inline lib/cylc/gui/app_gcylc.py]: fixed

**Comment by hjoliver** [comment]
This is (hopefully) good to go now.

**Pull request merged**